src/v/redpanda/admin_server.cc

```diff
@@ -212,21 +212,40 @@ void admin_server::log_level_timer_handler() {
     rearm_log_level_timer();
 }
 
-ss::future<ss::httpd::redirect_exception>
-admin_server::redirect_to_leader(ss::httpd::request& req) const {
-    auto leader_id_opt = _metadata_cache.local().get_leader_id(
-      model::controller_ntp);
+ss::future<ss::httpd::redirect_exception> admin_server::redirect_to_leader(
+  ss::httpd::request& req, model::ntp const& ntp) const {
+    auto leader_id_opt = _metadata_cache.local().get_leader_id(ntp);
 
     if (!leader_id_opt.has_value()) {
+        vlog(
+          logger.info,
+          "Can't redirect to leader, no leader for ntp {}",
+          leader_id_opt.value(),
+          ntp);
+
         throw ss::httpd::base_exception(
-          fmt::format("Controller leader not available"),
+          fmt::format(
+            "Partition {} does not have a leader, cannot redirect", ntp),
+          ss::httpd::reply::status_type::service_unavailable);
+    }
+
+    if (leader_id_opt.value() == config::node().node_id()) {
+        vlog(
+          logger.info,
+          "Can't redirect to leader from leader node ({})",
+          leader_id_opt.value());
+        throw ss::httpd::base_exception(
+          fmt::format("Leader not available"),
```

Comment:
It seems like in the generalization we'd retry the operation in this case, but to make that happen i think would require a lot of code changes. Rather, should we not say 'leader not available' but instead return some sort of retry code / message?

Reply:
The 503 we're returning should generally be interpreted by clients as a retryable status.

This particular case should be super rare, only coming up when leadership happens to transfer to the current node between the start of request processing and the point where we're generating a redirect.  Or some logic bug of course.  In the latter case it's more robust to finish the request and have the client retry than to have the request handler loop.